### ISSUE TITLE
react: Server side rendering can break when you use code that yields (node-fibers)

### ISSUE THREAD

**Issue by macrozone** [meteor/meteor#10364]
This is similar to this https://github.com/meteor/meteor/issues/10279

doing server side rendering (`ReactDom.renderToString`) with meteor and react >= 16.5 can throw an error:

```
Cannot read property 'debugElementStack' of undefined
```

This happens if two renderes start at the same time and if your App contains code that yield because of fibers.

## Background

In ReactDom server we have this piece of code:

```
pushCurrentDebugStack(this.stack);
try {
    out += this.render(child, frame.context, frame.domNamespace);
    } finally {
    popCurrentDebugStack();
}
```

`pushCurrentDebugStack` and `popCurrentDebugStack` both operate on a global array.

In a normal JS environment, this is no problem, because the three statements above always happen in this order:

// first render
`pushCurrentDebugStack`--> `this.render`--> `popCurrentDebugStack`
// second render
`pushCurrentDebugStack`--> `this.render`--> `popCurrentDebugStack`

If you have co-routines like fiber however, and in particular, if you use Collections or other fiber-yielding functions during `this.render`, `this.render` itself can yield and the order is no longer guaranteed, because the second render might start before the first finishes. This will lead to inconsistency.

## But isn't that a reactdom problem that should be fixed there?

Well... relying on the execution order like that is usually fine. They could fix it, by check which object to push and pop, instead relying on the order of execution. Could be an easy fix. But why should they fix it? They say:

> React supports standard-compliant JavaScript environments, so using an environment with non-compliant concurrency primitives is going to cause problems.

That's bad news for us, because it basically says that react does not support environments like meteor where we can have co-routines.

## Who is affected by that?

You will only run into this problem if 

- you render your app on the server (SSR, obviously)
- you use meteor-collections inside your render-functions (your `<App />`) or other fiber-yielding functions

So people using graphql/apollo or similar are not affected, because they have usually no yielding functions in their react-tree.

## Ok, the above error seems only to happen on develop. should i care?

This is not the only place where node-fibers and react-dom-server do not play nice together, see also this https://github.com/meteor/meteor/issues/10279

And we might have more errors like that in the future as long as meteor uses node-fibers or as long as react does not support such environments.

## What should we do?

i think we can't get rid of fibers easily, also in my opinion, it was a good tool when designing the old data-layer around pub/sub. It's also worth noting that SSR with meteor works quite well in general.

So our only chance is to get react to support it better, but that will also need some effort from the core-maintainers of meteor. At least in a way, so that we can monkey-patching situations like that (the fix usually is to attach globals to `require('fibers').current).

**Comment by CaptainN** [comment]
Is this still an issue, and does [this package](https://github.com/panter/meteor-fiber-save-react-context) still try to patch it?

**Comment by macrozone** [comment]
@CaptainN no, the issue has been solved in react 16.6.3.

**Comment by CaptainN** [comment]
I'm tacking a memory leak in my SSR code, which uses hooks and Mongo queries. I'm getting an error right before the server crashes, and the root of the stack trace is from a fiber pool file. The specific error is about using a react hook outside of a component, even though the component is listed in the stack trace. A local heap dump of the same code base shows a lot of allocations coming from the mongo driver. I'm wondering if there is a comparability problem between some mix of SSR, hooks, mongo and fibers. Maybe I should open a separate ticket for that.

**Comment by minhna** [comment]
@CaptainN I have problem with hooks and SSR too.
I use this kind of code to load the initial state data:
```
const initialState = {
    item: null,
    loading: Meteor.isClient,
    loaded: false,
  };

  if (Meteor.isServer) {
    initialState.item = Meteor.call('some.method', { itemId });
    initialState.loaded = true;
  }

  const reducer = (state, action) => { .... };

  const [state, dispatch] = useReducer(reducer, initialState);
```
The SSR `renderToString` function throws error when I try to hit f5 buttton fast enough. But after I add the `this.unblock()` to the method function, it's lots better.

**Comment by CaptainN** [comment]
My current suspicion is that while this may have been fixed in class based components, it is still a problem in function components or with hooks. React team warns against side effects in render, and this is technically a side effect.

**Comment by eugle** [comment]
The problem is still there

1. React 16.13.1
2. Meteor 1.10.2

```
I20200601-09:21:10.637(8)? TypeError: Cannot read property 'debugElementStack' of undefined
I20200601-09:21:10.637(8)?     at pushElementToDebugStack (/Users/mozibrand/Documents/git/me/jimuku/node_modules/react-dom/cjs/react-dom-server.node.development.js:2677:11)
I20200601-09:21:10.637(8)?     at resolve (/Users/mozibrand/Documents/git/me/jimuku/node_modules/react-dom/cjs/react-dom-server.node.development.js:2953:7)
I20200601-09:21:10.638(8)?     at ReactDOMServerRenderer.render (/Users/mozibrand/Documents/git/me/jimuku/node_modules/react-dom/cjs/react-dom-server.node.development.js:3435:22)
I20200601-09:21:10.638(8)?     at ReactDOMServerRenderer.read (/Users/mozibrand/Documents/git/me/jimuku/node_modules/react-dom/cjs/react-dom-server.node.development.js:3373:29)
I20200601-09:21:10.638(8)?     at renderToString (/Users/mozibrand/Documents/git/me/jimuku/node_modules/react-dom/cjs/react-dom-server.node.development.js:3988:27)
I20200601-09:21:10.638(8)?     at process (/Users/mozibrand/Documents/git/me/jimuku/node_modules/@apollo/react-ssr/lib/react-ssr.cjs.js:36:16)
I20200601-09:21:10.638(8)?     at /Users/mozibrand/.meteor/packages/promise/.0.11.2.gh9fq7.gg3jw++os+web.browser+web.browser.legacy+web.cordova/npm/node_modules/meteor-promise/fiber_pool.js:43:40
I20200601-09:21:10.638(8)?  => awaited here:
I20200601-09:21:10.639(8)?     at Function.Promise.await (/Users/mozibrand/.meteor/packages/promise/.0.11.2.gh9fq7.gg3jw++os+web.browser+web.browser.legacy+web.cordova/npm/node_modules/meteor-promise/promise_server.js:56:12)
I20200601-09:21:10.639(8)?     at imports/startup/server/ssr.js:48:16
I20200601-09:21:10.639(8)?     at /Users/mozibrand/.meteor/packages/promise/.0.11.2.gh9fq7.gg3jw++os+web.browser+web.browser.legacy+web.cordova/npm/node_modules/meteor-promise/fiber_pool.js:43:40
```

**Comment by CaptainN** [comment]
I wonder if something like [desync](https://github.com/abbr/deasync) could help?

**Comment by macrozone** [comment]
> I wonder if something like [desync](https://github.com/abbr/deasync) could help?

that would make it even worse ;-)

The problem is if you write a library like react and users of your libraries can provide a `render` function, you might have something like this in your library:

```
// A
const result = userProvidedRenderFunction(props) // B
// C
```

The language spec makes it clear that A, B, C run in that order on the one node process.
So the library can use some globals or shared variables in A and C and can assume that nothing happens between and can assume that these variables won't be altered unexpectedly. So if multiple components are rendered, the order will always be A,B,C, A,B,C, etc.

This is NOT the case with fibers or desync. The `userProvidedRenderFunction` could yield therefore can alter the execution order. 

So for example if you have multiple components, the execution order can change: 
A, B starts but yields, A from 2 component, B from second components, B from first resumes, C from first, B from second resumes, C from second. 

As a simple example that breaks with fiber, take `console.time(label) / console.timeEnd(label)`:

```
console.time("render duration") // A
const result = userProvidedRenderFunction(props) // B
console.timeEnd("render duration")// C
```

in spec-compliant js, this always works, but not with fibers.


This is also one reason why you have to annotate async functions in ES (With `async`, or generator functions with `*`). it makes it for all very obvious where the order of statements is in strict order (--> if there is no `await`)

**Comment by CaptainN** [comment]
@macrozone If that's the case then using Meteor Mongo in SSR with React should be specifically prohibited...

**Comment by macrozone** [comment]
> @macrozone If that's the case then using Meteor Mongo in SSR with React should be specifically prohibited...

this is not only a react problem, it _could_ apply to _every library_. This does not mean that it doesn't work at all. In fact, SSR with meteor, react and fiber-driven mongo collections _does_ work (we have one last production app using that). It just can break with some version under specific conditions (as it already did). it is like a grey zone and that makes me uncomfortable to use it.

**Comment by CaptainN** [comment]
Couple of ideas, observations first:
- The output works - even though I get warnings, my HTML output is still what I expect.
- It leaks memory - I haven't been able to gain any insight as to why yet except that a fiber pool is reference at the base of the "out of memory" error. I need to learn how to get some measurements going (I host on Galaxy, and don't do dev ops much).
- SSR passes are a relatively constrained process - going from start to finish

Without measurements, I'm probably spinning wheels, but here's an idea to work around the problem:
- Parse the react tree without rendering anything (there are examples of how to do this in Apollo's SSR code). At this time extract the queries and other fiber/async requests (methods), but don't run them.
- Run the Queries and cache the results.
- (Maybe run the tree parser again, to capture derivative queries, recurse until no new queries are added to the queue/cache)
- Render to string (or to nodeStream) using the cached query data. No need 

Something similar could probably be set up for non-React SSR too (Svelte, Vue).

**Comment by CaptainN** [comment]
@macrozone Does your SSR code use the newer `useTracker` (or `withTracker` built on `useTracker`)? I'm thinking these leaks are specifically from an interplay between react's hooks and Meteor's Fibers.

**Comment by onzag** [comment]
I am getting the same issue with my own framework, there's a fallback in place for when this happens, the client is instructed to not use SSR; I only managed to reproduce it when testing my server into very bad connection scenarios.

Is there a way to prevent this issue altogether?

**Comment by macrozone** [comment]
> @macrozone Does your SSR code use the newer `useTracker` (or `withTracker` built on `useTracker`)? I'm thinking these leaks are specifically from an interplay between react's hooks and Meteor's Fibers.

sorry for the late response, we did not use hooks in that project and we also did not have leaks.
